### PR TITLE
New module - elb_target

### PR DESCRIPTION
**Pull request by wimnat**
##### SUMMARY
New module - elb_target

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
elb_target

##### ANSIBLE VERSION
```
2.5
```

##### ADDITIONAL INFORMATION
@willthames @s-hertel

### CODE REVIEW

**Comment by wimnat** [comment]
@jbscare

File: lib/ansible/modules/cloud/amazon/elb_target.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/elb_target.py]: If the TargetHealthDescriptions are an empty list this will fail. Maybe check if it exists and if not return an empty dict?
 Traceback:
```
The full traceback is:
Traceback (most recent call last):
  File "/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_mqh88b9h/ansible_module_elb_target.py", line 245, in <module>
    main()
  File "/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_mqh88b9h/ansible_module_elb_target.py", line 242, in main
    deregister_target(connection, module)
  File "/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_mqh88b9h/ansible_module_elb_target.py", line 205, in deregister_target
    target_descriptions = describe_targets(connection, module, target_group_arn, [])
  File "/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_mqh88b9h/ansible_module_elb_target.py", line 125, in describe_targets
    return connection.describe_target_health(TargetGroupArn=tg_arn, Targets=target)['TargetHealthDescriptions'][0]
IndexError: list index out of range

fatal: [localhost]: FAILED! => {
    "changed": false,
    "failed": true,
    "module_stderr": "Traceback (most recent call last):\n  File \"/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_mqh88b9h/ansible_module_elb_target.py\", line 245, in <module>\n    main()\n  File \"/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_mqh88b9h/ansible_module_elb_target.py\", line 242, in main\n    deregister_target(connection, module)\n  File \"/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_mqh88b9h/ansible_module_elb_target.py\", line 205, in deregister_target\n    target_descriptions = describe_targets(connection, module, target_group_arn, [])\n  File \"/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_mqh88b9h/ansible_module_elb_target.py\", line 125, in describe_targets\n    return connection.describe_target_health(TargetGroupArn=tg_arn, Targets=target)['TargetHealthDescriptions'][0]\nIndexError: list index out of range\n",
    "module_stdout": "",
    "msg": "MODULE FAILURE",
    "rc": 0
}

PLAY RECAP *************************************************************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1
```

Comment (s-hertel) [reply lib/ansible/modules/cloud/amazon/elb_target.py]: Also, what if the target has been registered on multiple ports (since running the playbook with a new port doesn't deregister the target with the old port)? Should those be in the returned info as well, instead of just the first item in the list?

**Review by s-hertel — CHANGES_REQUESTED** [review]

**Review by s-hertel — APPROVED** [review]

**Comment by kernow** [comment]
@wimnat I've been using this module and encountered an issue with the way you are handling existing targets, attempting to register or deregister an existing target causes an exception. In both `register_target` and `deregister_target` the code relies on the `Reason` being supplied in the `TargetHealth` object. This fails when the targets health state is `healthy` as the `Reason` is not supplied by the api when the state is `healthy`. See: [TargetHealth API docs](http://docs.aws.amazon.com/elasticloadbalancing/latest/APIReference/API_TargetHealth.html)

As the `State` may or may not be returned I believe you will first need to check if the `State` is present, if it's not then check if a `Reason` is present, then check for the value of the `Reason`

**Comment by kernow** [comment]
@wimnat I wonder if this would work for all situations for registration:

    if 'Reason' in target_description['TargetHealth']:
        if target_description['TargetHealth']['Reason'] == "Target.NotRegistered":

and for deregistration:

    if 'State' in target_description['TargetHealth']:
        if target_description['TargetHealth']['State'] != 'draining':

**Comment by Lansoweb** [comment]
@kernow I used your suggestion today with no luck, the script just skipped the registration or deregistration.

**Comment by kernow** [comment]
@Lansoweb This is what I'm running and its working without issues for me https://gist.github.com/kernow/89d2e3609fa5ea2e75b33fea6b3b4b98

**Comment by jurajseffer** [comment]
I've tried to improve on this module by including parameters that make it wait for status change, useful during rolling deploys. https://gist.github.com/jurajseffer/20b8df6c423723f54ed411c740b06bc6

**Comment by wimnat** [comment]
@jurajseffer PRs to my branch welcome

**Comment by hylaride** [comment]
Super looking forward to this!

**Comment by hylaride** [comment]
@jurajseffer I'm using your improvements and they're great!  Please get a PR going! 😄

**Comment by willthames** [comment]
rebuild_merge

**Comment by willthames** [comment]
I'm happy enough to merge this once tests pass (it doesn't have to be perfect on day one for a `preview` module) and let any further improvements come through ansible/ansible PRs.

**Comment by kernow** [comment]
@willthames has the issue I raised been fixed? I don't think it should be merged until the issue is resolved as it causes an exception if you try and work with targets that are already registered. I've posted a solution that works for me so it would be easy to incorporate this

**Comment by s-hertel** [comment]
@kernow I think I fixed it with 1df6996. Will you test and see if that's working right for you?
I'm still checking 'Reason' in target_description['TargetHealth'] for deregistering (because the target health status could be `unused` which can correlate to too many things - an unregistered target or that the target group is not used in a listener rule for the load balancer, or that the target in an AZ that wasn't enabled for the load balancer) but I've also taken into account that no reason means the target can be deregistered.

**Comment by wimnat** [comment]
@s-hertel after looking at both sets of changes I decided to go with @jurajseffer changes (sorry!)

@kernow can you please retest with latest changes?

I'm removing the WIP from this PR with a view to merging for 2.5

**Comment by ocervell** [comment]
Can we have this merged to master ? It's critical for us to have this feature to remove any manual work.

**Comment by s-hertel** [comment]
@ocervell Definitely looking forward to having this in 2.5. Community module freeze is at the end of January, so you can plan on seeing this merged before then.

@wimnat Cool, @jurajseffer's changes look good. The bug I made note of in https://github.com/ansible/ansible/commit/1df699668d93ffb33f24ad1ec1d7968943215641 still exists, so it would be nice to see that fixed. Here's a minimal reproducer to use that code path:
```
---
- hosts: localhost
  connection: local
  gather_facts: no
  vars:
    name: "{{ target_group }}"
    port: "{{ target_port }}"
    target_id: "{{ target_id }}"
  tasks:
    - name: register a target without configuring to receive traffic from the load balancer
      elb_target:
        target_group_name: "{{ name }}"
        target_id: "{{ target_id }}"
        target_port: "{{ port }}"
        state: present

    - name: attempt to deregister an unused target
      elb_target:
        target_group_name: "{{ name }}"
        target_id: "{{ target_id }}"
        state: absent
```
The behavior:
The first task reports changed: true. The subsequent task and all subsequent playbook runs are changed: false.
The expected behavior:
To be able to remove unused targets. Rather than appearing to already having been deleted and succeeding silently with changed: false, an error to indicate it isn't actually being deleted seems like more transparent behavior. But it's entirely possible to remove unused targets, so that behavior would ideally be supported. The code resulting in this behavior is [here](https://github.com/ansible/ansible/pull/26483/files#diff-30dcc67612416d19eaf6489d4f424ddaR207).

**Comment by wimnat** [comment]
@s-hertel does an Ansible module have a way to print a warning message?

I can add a parameter so that unused targets can be removed but I think that you're right in that we want to tell a user when a target still exists because it is unused.  My preferred method would be a warning message but I don't think I've ever seen Ansible do this :/

**Comment by jurajseffer** [comment]
@wimnat This module seems to return an object with warnings in it https://github.com/ansible/ansible/blob/9f86b923e96f58c33efcca6d19ab01c166a226b9/lib/ansible/modules/network/nxos/nxos_nxapi.py#L289

**Comment by s-hertel** [comment]
@wimnat You can use `module.warn("Your warning here")`.

**Comment by wimnat** [comment]
@s-hertel - i've added a new option `deregister_unused` to force deregistration of targets otherwise it'll give a warning that it is still registered but unused.

I've also added a link to more AWS documentation on target status.

@kernow - can you confirm the latest changes work for you?

File: lib/ansible/modules/cloud/amazon/elb_target.py
Comment (rdeknijf) [inline lib/ansible/modules/cloud/amazon/elb_target.py]: This module only deregisters a module if `current_target_state` is `unused`. I would like to use it to deregister used targets. This fails silently now.

Comment (jurajseffer) [reply lib/ansible/modules/cloud/amazon/elb_target.py]: Yes, from what I can see this doesn't work now. My `state: absent` is ignored when target is used by target group. Where is `else`? :)

Comment (tedder) [reply lib/ansible/modules/cloud/amazon/elb_target.py]: yeah, there's a missing chunk of code.

**Comment by tedder** [comment]
@s-hertel @wimnat note the code comments above- it seems like one of the major code paths (unregistering healthy hosts) simply doesn't work. I've duplicated code locally to fix it, but I'm missing something on assumptions in that method.

**Comment by jurajseffer** [comment]
I have another case that gave me an unexpected behaviour. I tried calling this module twice, once without that blocking wait for status change. e.g.
```
elb_target:
        region: "{{ ec2_region }}"
        target_group_arn: "{{ item }}"
        target_id: "{{ ec2_id }}"
        state: absent
with_items: some_arns
```
so that multiple target groups start to drain all at once. This worked (with an older version as the current version here in the PR doesn't work at all for this). Then I make another call which should be blocking, waiting for all target groups I'm iterating over. E.g.
```
elb_target:
        region: "{{ ec2_region }}"
        target_group_arn: "{{ item }}"
        target_id: "{{ ec2_id }}"
        state: absent
        target_status: unused
with_items: some_arns
```
This reported as `OK` but the state of the instance was still `draining` on the target group so perhaps it only checked that it's already draining because of `absent` state without checking that it's unused.
I guess this check cannot be a subject to the above condition where we check whether it's already in draining or unused state
https://github.com/ansible/ansible/pull/26483/files#diff-30dcc67612416d19eaf6489d4f424ddaR236

**Comment by s-hertel** [comment]
@wimnat @tedder @jurajseffer I've added tests and believe I have fixed the deregistering bugs. I haven't added the alias to run in CI (which would need to be updated with IAM permissions and I'll work on that after this has been merged since I don't want to hold it up more). Does this look okay to everyone? If you have time, please re-test.

**Comment by s-hertel** [comment]
bot_status

**Comment by s-hertel** [comment]
!needs_revision

**Comment by tedder** [comment]
testing locally.

**Comment by tedder** [comment]
fixes my objection. `shipit`.

**Comment by s-hertel** [comment]
The bot was needs revision because there was an @ in a commit message. I rebased to remove that from the commit message.

**Review by ryansb — APPROVED** [review]

**Pull request merged**
